File: webflux-dialogue/src/test/java/com/study/webflux/rag/infrastructure/adapter/vectordb/SpringAiVectorDbAdapterTest.java
package com.study.webflux.rag.infrastructure.adapter.vectordb;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyList;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import java.time.Instant;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.ai.document.Document;
import org.springframework.ai.vectorstore.SearchRequest;
import org.springframework.ai.vectorstore.VectorStore;

import com.study.webflux.rag.domain.model.memory.Memory;
import com.study.webflux.rag.domain.model.memory.MemoryType;

import reactor.test.StepVerifier;

@ExtendWith(MockitoExtension.class)
class SpringAiVectorDbAdapterTest {

	@Mock
	private VectorStore vectorStore;

	private SpringAiVectorDbAdapter vectorDbAdapter;

	@BeforeEach
	void setUp() {
		vectorDbAdapter = new SpringAiVectorDbAdapter(vectorStore);
	}

	@Test
	@DisplayName("메모리를 벡터 DB에 저장한다")
	void upsert_success() {
		Memory memory = new Memory(
			null,
			MemoryType.EXPERIENTIAL,
			"테스트 메모리",
			0.8f,
			Instant.now(),
			Instant.now(),
			1
		);
		List<Float> embedding = List.of(0.1f, 0.2f, 0.3f);

		StepVerifier.create(vectorDbAdapter.upsert(memory, embedding))
			.assertNext(result -> {
				assertThat(result).isNotNull();
				assertThat(result.id()).isNotNull();
				assertThat(result.content()).isEqualTo("테스트 메모리");
				assertThat(result.type()).isEqualTo(MemoryType.EXPERIENTIAL);
			})
			.verifyComplete();

		verify(vectorStore).add(anyList());
	}

	@Test
	@DisplayName("ID가 있는 메모리를 저장할 때 ID를 유지한다")
	void upsert_withExistingId_preservesId() {
		String existingId = "existing-id-123";
		Memory memory = new Memory(
			existingId,
			MemoryType.FACTUAL,
			"기존 메모리",
			0.5f,
			null,
			null,
			null
		);
		List<Float> embedding = List.of(0.1f, 0.2f);

		StepVerifier.create(vectorDbAdapter.upsert(memory, embedding))
			.assertNext(result -> {
				assertThat(result.id()).isEqualTo(existingId);
			})
			.verifyComplete();
	}

	@Test
	@DisplayName("메모리 저장 시 메타데이터를 올바르게 설정한다")
	void upsert_setsMetadataCorrectly() {
		Instant createdAt = Instant.parse("2025-01-01T00:00:00Z");
		Instant lastAccessedAt = Instant.parse("2025-01-02T00:00:00Z");

		Memory memory = new Memory(
			null,
			MemoryType.EXPERIENTIAL,
			"메타데이터 테스트",
			0.9f,
			createdAt,
			lastAccessedAt,
			5
		);
		List<Float> embedding = List.of(0.1f);

		ArgumentCaptor<List<Document>> captor = ArgumentCaptor.forClass(List.class);

		StepVerifier.create(vectorDbAdapter.upsert(memory, embedding))
			.assertNext(result -> assertThat(result).isNotNull())
			.verifyComplete();

		verify(vectorStore).add(captor.capture());
		Document capturedDoc = captor.getValue().get(0);
		Map<String, Object> metadata = capturedDoc.getMetadata();

		assertThat(metadata.get("type")).isEqualTo("EXPERIENTIAL");
		assertThat(metadata.get("importance")).isEqualTo(0.9f);
		assertThat(metadata.get("createdAt")).isEqualTo(createdAt.getEpochSecond());
		assertThat(metadata.get("lastAccessedAt")).isEqualTo(lastAccessedAt.getEpochSecond());
		assertThat(metadata.get("accessCount")).isEqualTo(5);
	}

	@Test
	@DisplayName("타입과 중요도 필터로 메모리를 검색한다")
	void search_withFilters_success() {
		List<Float> queryEmbedding = List.of(0.1f, 0.2f, 0.3f);
		List<MemoryType> types = List.of(MemoryType.EXPERIENTIAL, MemoryType.FACTUAL);
		float importanceThreshold = 0.5f;
		int topK = 5;

		Map<String, Object> metadata = new HashMap<>();
		metadata.put("type", "EXPERIENTIAL");
		metadata.put("importance", 0.8);
		metadata.put("createdAt", Instant.now().getEpochSecond());

		Document mockDocument = new Document("doc-1", "테스트 내용", metadata);
		List<Document> mockResults = List.of(mockDocument);

		when(vectorStore.similaritySearch(any(SearchRequest.class))).thenReturn(mockResults);

		StepVerifier.create(vectorDbAdapter.search(queryEmbedding, types, importanceThreshold, topK))
			.assertNext(result -> {
				assertThat(result).isNotNull();
				assertThat(result.id()).isEqualTo("doc-1");
				assertThat(result.content()).isEqualTo("테스트 내용");
				assertThat(result.type()).isEqualTo(MemoryType.EXPERIENTIAL);
				assertThat(result.importance()).isEqualTo(0.8f);
			})
			.verifyComplete();

		ArgumentCaptor<SearchRequest> captor = ArgumentCaptor.forClass(SearchRequest.class);
		verify(vectorStore).similaritySearch(captor.capture());

		SearchRequest capturedRequest = captor.getValue();
		assertThat(capturedRequest.getTopK()).isEqualTo(topK);
		String expressionDescription = String.valueOf(capturedRequest.getFilterExpression());
		assertThat(expressionDescription).contains("Value[value=[EXPERIENTIAL, FACTUAL]]");
		assertThat(expressionDescription).contains("Key[key=importance]");
		assertThat(expressionDescription).contains("Value[value=0.5]");
	}

	@Test
	@DisplayName("검색 결과가 없으면 빈 Flux를 반환한다")
	void search_noResults_returnsEmpty() {
		List<Float> queryEmbedding = List.of(0.1f);
		List<MemoryType> types = List.of(MemoryType.EXPERIENTIAL);

		when(vectorStore.similaritySearch(any(SearchRequest.class))).thenReturn(List.of());

		StepVerifier.create(vectorDbAdapter.search(queryEmbedding, types, 0.5f, 5))
			.verifyComplete();
	}

	@Test
	@DisplayName("여러 메모리를 검색하여 반환한다")
	void search_multipleResults_returnsAll() {
		List<Float> queryEmbedding = List.of(0.1f);
		List<MemoryType> types = List.of(MemoryType.EXPERIENTIAL);

		Map<String, Object> metadata1 = new HashMap<>();
		metadata1.put("type", "EXPERIENTIAL");
		metadata1.put("importance", 0.9);

		Map<String, Object> metadata2 = new HashMap<>();
		metadata2.put("type", "EXPERIENTIAL");
		metadata2.put("importance", 0.7);

		Document doc1 = new Document("doc-1", "첫 번째 메모리", metadata1);
		Document doc2 = new Document("doc-2", "두 번째 메모리", metadata2);

		when(vectorStore.similaritySearch(any(SearchRequest.class)))
			.thenReturn(List.of(doc1, doc2));

		StepVerifier.create(vectorDbAdapter.search(queryEmbedding, types, 0.5f, 5))
			.assertNext(result -> {
				assertThat(result.id()).isEqualTo("doc-1");
				assertThat(result.importance()).isEqualTo(0.9f);
			})
			.assertNext(result -> {
				assertThat(result.id()).isEqualTo("doc-2");
				assertThat(result.importance()).isEqualTo(0.7f);
			})
			.verifyComplete();
	}

	@Test
	@DisplayName("중요도 업데이트는 로그만 출력한다")
	void updateImportance_logsWarning() {
		String memoryId = "test-id";
		float newImportance = 0.9f;
		Instant lastAccessedAt = Instant.now();
		int accessCount = 10;

		StepVerifier.create(
			vectorDbAdapter.updateImportance(memoryId, newImportance, lastAccessedAt, accessCount)
		).verifyComplete();
	}

	@Test
	@DisplayName("Document에서 Memory로 변환 시 모든 필드를 매핑한다")
	void search_mapsAllFields() {
		Instant now = Instant.now();

		Map<String, Object> metadata = new HashMap<>();
		metadata.put("type", "FACTUAL");
		metadata.put("importance", 0.85);
		metadata.put("createdAt", now.minusSeconds(3600).getEpochSecond());
		metadata.put("lastAccessedAt", now.getEpochSecond());
		metadata.put("accessCount", 3);

		Document document = new Document("doc-123", "완전한 메모리", metadata);

		when(vectorStore.similaritySearch(any(SearchRequest.class)))
			.thenReturn(List.of(document));

		StepVerifier.create(
			vectorDbAdapter.search(List.of(0.1f), List.of(MemoryType.FACTUAL), 0.5f, 1)
		)
			.assertNext(result -> {
				assertThat(result.id()).isEqualTo("doc-123");
				assertThat(result.content()).isEqualTo("완전한 메모리");
				assertThat(result.type()).isEqualTo(MemoryType.FACTUAL);
				assertThat(result.importance()).isEqualTo(0.85f);
				assertThat(result.createdAt()).isNotNull();
				assertThat(result.lastAccessedAt()).isNotNull();
				assertThat(result.accessCount()).isEqualTo(3);
			})
			.verifyComplete();
	}

	@Test
	@DisplayName("메타데이터에 선택적 필드가 없어도 정상 처리한다")
	void search_withMissingOptionalFields_success() {
		Map<String, Object> metadata = new HashMap<>();
		metadata.put("type", "EXPERIENTIAL");

		Document document = new Document("doc-min", "최소 메모리", metadata);

		when(vectorStore.similaritySearch(any(SearchRequest.class)))
			.thenReturn(List.of(document));

		StepVerifier.create(
			vectorDbAdapter.search(List.of(0.1f), List.of(MemoryType.EXPERIENTIAL), 0.0f, 1)
		)
			.assertNext(result -> {
				assertThat(result.id()).isEqualTo("doc-min");
				assertThat(result.type()).isEqualTo(MemoryType.EXPERIENTIAL);
				assertThat(result.importance()).isNull();
				assertThat(result.createdAt()).isNull();
				assertThat(result.lastAccessedAt()).isNull();
				assertThat(result.accessCount()).isNull();
			})
			.verifyComplete();
	}

	@Test
	@DisplayName("단일 타입으로 검색 시 필터 표현식이 올바르게 생성된다")
	void search_singleType_createsCorrectFilter() {
		List<MemoryType> types = List.of(MemoryType.FACTUAL);

		when(vectorStore.similaritySearch(any(SearchRequest.class))).thenReturn(List.of());

		StepVerifier.create(vectorDbAdapter.search(List.of(0.1f), types, 0.3f, 5))
			.verifyComplete();

		ArgumentCaptor<SearchRequest> captor = ArgumentCaptor.forClass(SearchRequest.class);
		verify(vectorStore).similaritySearch(captor.capture());

		String filterExpression = String.valueOf(captor.getValue().getFilterExpression());
		assertThat(filterExpression).contains("Value[value=[FACTUAL]]");
		assertThat(filterExpression).contains("Key[key=importance]");
		assertThat(filterExpression).contains("Value[value=0.3]");
	}
}

File: app-miyou/src/test/kotlin/com/miyou/app/infrastructure/memory/adapter/SpringAiVectorDbAdapterTest.kt
package com.miyou.app.infrastructure.memory.adapter

import com.google.common.util.concurrent.Futures
import com.miyou.app.domain.memory.model.Memory
import com.miyou.app.domain.memory.model.MemoryEmotion
import com.miyou.app.domain.memory.model.MemoryType
import com.miyou.app.fixture.ConversationSessionFixture
import com.miyou.app.infrastructure.dialogue.config.properties.RagDialogueProperties
import com.miyou.app.support.anyValue
import com.miyou.app.support.eqValue
import io.qdrant.client.QdrantClient
import io.qdrant.client.grpc.JsonWithInt
import io.qdrant.client.grpc.Points
import io.qdrant.client.grpc.Points.ScoredPoint
import io.qdrant.client.grpc.Points.ScrollPoints
import io.qdrant.client.grpc.Points.ScrollResponse
import io.qdrant.client.grpc.Points.SearchPoints
import io.qdrant.client.grpc.Points.UpdateResult
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.ArgumentCaptor
import org.mockito.ArgumentMatchers.any
import org.mockito.ArgumentMatchers.nullable
import org.mockito.Mock
import org.mockito.Mockito.verify
import org.mockito.Mockito.`when`
import org.mockito.junit.jupiter.MockitoExtension
import org.springframework.ai.document.Document
import org.springframework.ai.vectorstore.VectorStore
import reactor.test.StepVerifier
import java.time.Instant

@ExtendWith(MockitoExtension::class)
class SpringAiVectorDbAdapterTest {
    @Mock
    private lateinit var vectorStore: VectorStore

    @Mock
    private lateinit var qdrantClient: QdrantClient

    private lateinit var vectorDbAdapter: SpringAiVectorDbAdapter

    @BeforeEach
    fun setUp() {
        val properties = RagDialogueProperties()
        properties.qdrant.collectionName = "test-collection"
        vectorDbAdapter = SpringAiVectorDbAdapter(vectorStore, qdrantClient, properties)
    }

    @Test
    @DisplayName("upsert 시 메모리 메타데이터를 벡터 저장소에 저장한다")
    fun upsert_storesMemoryMetadata() {
        val sessionId = ConversationSessionFixture.createId()
        val memory =
            Memory(
                id = null,
                sessionId = sessionId,
                type = MemoryType.EXPERIENTIAL,
                content = "metadata test",
                importance = 0.9f,
                createdAt = Instant.parse("2025-01-01T00:00:00Z"),
                lastAccessedAt = Instant.parse("2025-01-02T00:00:00Z"),
                accessCount = 5,
                emotion = MemoryEmotion.POSITIVE,
            )

        @Suppress("UNCHECKED_CAST")
        val captor = ArgumentCaptor.forClass(List::class.java) as ArgumentCaptor<List<Document>>

        StepVerifier
            .create(vectorDbAdapter.upsert(memory, listOf(0.1f)))
            .assertNext { result -> assertThat(result.id).isNotNull() }
            .verifyComplete()

        verify(vectorStore).add(captor.capture())
        val metadata = captor.value.single().metadata
        assertThat(metadata["sessionId"]).isEqualTo(sessionId.value)
        assertThat(metadata["type"]).isEqualTo("EXPERIENTIAL")
        assertThat(metadata["importance"]).isEqualTo(0.9f)
        assertThat(metadata["accessCount"]).isEqualTo(5)
        assertThat(metadata["emotion"]).isEqualTo("POSITIVE")
    }

    @Test
    @DisplayName("검색 결과 Qdrant 포인트를 Memory 객체로 변환한다")
    fun search_mapsQdrantPointsIntoMemoryObjects() {
        val sessionId = ConversationSessionFixture.createId()
        val point =
            ScoredPoint
                .newBuilder()
                .setId(
                    Points.PointId
                        .newBuilder()
                        .setUuid("doc-1")
                        .build(),
                ).putPayload(
                    "doc_content",
                    JsonWithInt.Value
                        .newBuilder()
                        .setStringValue("test content")
                        .build(),
                ).putPayload(
                    "type",
                    JsonWithInt.Value
                        .newBuilder()
                        .setStringValue("EXPERIENTIAL")
                        .build(),
                ).putPayload(
                    "importance",
                    JsonWithInt.Value
                        .newBuilder()
                        .setDoubleValue(0.8)
                        .build(),
                ).build()

        `when`(qdrantClient.searchAsync(any(SearchPoints::class.java)))
            .thenReturn(Futures.immediateFuture(listOf(point)))

        StepVerifier
            .create(
                vectorDbAdapter.search(
                    sessionId,
                    listOf(0.1f, 0.2f),
                    listOf(MemoryType.EXPERIENTIAL),
                    0.5f,
                    5,
                ),
            ).assertNext { result ->
                assertThat(result.id).isEqualTo("doc-1")
                assertThat(result.content).isEqualTo("test content")
                assertThat(result.type).isEqualTo(MemoryType.EXPERIENTIAL)
            }.verifyComplete()
    }

    @Test
    @DisplayName("payload의 emotion 문자열을 MemoryEmotion으로 역직렬화한다")
    fun search_mapsEmotionFieldFromPayload() {
        val sessionId = ConversationSessionFixture.createId()
        val point =
            ScoredPoint
                .newBuilder()
                .setId(
                    Points.PointId
                        .newBuilder()
                        .setUuid("doc-2")
                        .build()
                ).putPayload(
                    "doc_content",
                    JsonWithInt.Value
                        .newBuilder()
                        .setStringValue("shocking content")
                        .build()
                ).putPayload(
                    "type",
                    JsonWithInt.Value
                        .newBuilder()
                        .setStringValue("EXPERIENTIAL")
                        .build()
                ).putPayload(
                    "importance",
                    JsonWithInt.Value
                        .newBuilder()
                        .setDoubleValue(0.4)
                        .build()
                ).putPayload(
                    "emotion",
                    JsonWithInt.Value
                        .newBuilder()
                        .setStringValue("SHOCKING")
                        .build()
                ).build()

        `when`(qdrantClient.searchAsync(any(SearchPoints::class.java)))
            .thenReturn(Futures.immediateFuture(listOf(point)))

        StepVerifier
            .create(
                vectorDbAdapter.search(sessionId, listOf(0.1f, 0.2f), listOf(MemoryType.EXPERIENTIAL), 0.5f, 5),
            ).assertNext { result -> assertThat(result.emotion).isEqualTo(MemoryEmotion.SHOCKING) }
            .verifyComplete()
    }

    @Test
    @DisplayName("updateImportance는 Qdrant payload를 갱신한다")
    fun updateImportance_writesPayloadToQdrant() {
        val memoryId = "550e8400-e29b-41d4-a716-446655440000"
        `when`(
            qdrantClient.setPayloadAsync(
                anyValue<String>(),
                anyValue<Map<String, JsonWithInt.Value>>(),
                anyValue<List<Points.PointId>>(),
                eqValue(true),
                nullable(Points.WriteOrderingType::class.java),
                nullable(java.time.Duration::class.java),
            ),
        ).thenReturn(Futures.immediateFuture(UpdateResult.getDefaultInstance()))

        StepVerifier
            .create(vectorDbAdapter.updateImportance(memoryId, 0.9f, Instant.now(), 10))
            .verifyComplete()
    }

    @Test
    @DisplayName("findAllActive는 archivedAt이 없는 포인트만 스크롤로 조회한다")
    fun findAllActive_scrollsOnlyNonArchivedPoints() {
        val point =
            Points.RetrievedPoint
                .newBuilder()
                .setId(
                    Points.PointId
                        .newBuilder()
                        .setUuid("mem-1")
                        .build()
                ).putPayload(
                    "sessionId",
                    JsonWithInt.Value
                        .newBuilder()
                        .setStringValue("session-1")
                        .build()
                ).putPayload(
                    "doc_content",
                    JsonWithInt.Value
                        .newBuilder()
                        .setStringValue("content")
                        .build()
                ).putPayload(
                    "type",
                    JsonWithInt.Value
                        .newBuilder()
                        .setStringValue("FACTUAL")
                        .build()
                ).putPayload(
                    "importance",
                    JsonWithInt.Value
                        .newBuilder()
                        .setDoubleValue(0.2)
                        .build()
                ).build()
        val response = ScrollResponse.newBuilder().addResult(point).build()

        `when`(qdrantClient.scrollAsync(any(ScrollPoints::class.java)))
            .thenReturn(Futures.immediateFuture(response))

        StepVerifier
            .create(vectorDbAdapter.findAllActive(100))
            .assertNext { memory ->
                assertThat(memory.id).isEqualTo("mem-1")
                assertThat(memory.sessionId.value).isEqualTo("session-1")
                assertThat(memory.archivedAt).isNull()
            }.verifyComplete()
    }

    @Test
    @DisplayName("applyDecayAndArchive는 importance와 archivedAt을 함께 반영한다")
    fun applyDecayAndArchive_writesImportanceAndArchivedAt() {
        val sessionId = ConversationSessionFixture.createId()
        val now = Instant.now()
        val archivedMemory =
            Memory(
                id = "550e8400-e29b-41d4-a716-446655440000",
                sessionId = sessionId,
                type = MemoryType.FACTUAL,
                content = "content",
                importance = 0.05f,
                createdAt = now,
                lastAccessedAt = now,
                accessCount = 1,
                archivedAt = now,
            )

        `when`(
            qdrantClient.setPayloadAsync(
                anyValue<String>(),
                anyValue<Map<String, JsonWithInt.Value>>(),
                anyValue<List<Points.PointId>>(),
                eqValue(true),
                nullable(Points.WriteOrderingType::class.java),
                nullable(java.time.Duration::class.java),
            ),
        ).thenReturn(Futures.immediateFuture(UpdateResult.getDefaultInstance()))

        StepVerifier
            .create(vectorDbAdapter.applyDecayAndArchive(archivedMemory))
            .verifyComplete()
    }
}
